File: Lesson4/Lesson_15/arrayTools.c
/*
 * arrayTools.c
 *
 *  Created on: 24 Apr 2020
 *      Author: titan
 */


#include<stdio.h>
#include<stdlib.h>
#include<time.h>


void printArr(int arr[], int size);
void fillArr(int arr[], int size, int min, int max);
void reverseArray(int arr[], int size);
int search(int arr[], int size, int value);
int maxArray(int arr[], int size);
int maxIndex(int arr[], int size);
int sumArray(int arr[], int size);
float arrayAvg(int arr[], int size);


int main () {

	printf("%d\n", RAND_MAX);

	int size = 10;
	int arr[size];

	fillArr(arr, size, 10, 99);

	printArr(arr, size);

	int res = search(arr, size, 55);
	printf("res = %d\n", res);
	if (res >= 0) {
		printf("arr[%d] = %d", res, arr[res]);
	}

	res = maxArray(arr, size);
	printf("max = %d\n", res);

	res = maxIndex(arr, size);
	printf("max Index = %d\n", res);

	res = sumArray(arr, size);
	printf("Array sum is %d\n", res);

	float avg = arrayAvg(arr, size);
	printf("Array average number is %.2f\n", avg);

	void reverseArray(arr, size);


	return 0;
}

void reverseArray(int arr[], int size) {
	int i;
	for (i = 0; i < size/2; i++) {
		int temp = arr[i];
		arr[i] = arr[size - 1 - i];
		arr[size - 1 - i] = temp;
	}

}

int sumArray(int arr[], int size) {
	int i;
	int sum = 0;
	for (i = 0; i < size; i++) {
		sum = sum + arr[i];
	}
	return sum;
}


float arrayAvg(int arr[], int size) {
//	float avg = (float)sumArray(arr, size) / size;
//	return avg;
	int i;
	float avg = 0;
	for (i = 0; i < size; i++) {
		avg = avg + arr[i]*1.0/size;
	}
	return avg;
}



int maxIndex(int arr[], int size){
	int i;
	int index = 0;
	for (i = 1; i < size; i++) {
		if (arr[index] < arr[i]) {
			index = i;
		}
	}
	return index;
//	int max = maxArray(arr, size);
//	return search(arr, size, max);
}

int maxArray(int arr[], int size) {
	int i;
	int max = arr[0];
	for (i = 1; i < size; i++) {
		if (max < arr[i]) {
			max = arr[i];
		}
	}
	return max;
}


void fillArr(int arr[], int size, int min, int max) {
	srand(time(NULL));
	int i;
	for (i = 0; i < size; i++) {
		arr[i] = min + rand() % (max + 1 - min);
	}
}


int search(int arr[], int size, int value) {
	int i;
	for (i = 0; i < size; i++) {
		if (arr[i] == value) {
			return i;
		}
	}
return -1;
}

void printArr(int arr[], int size) {
	int i;
	printf("{ ");
	for (i = 0; i < size; i++) {
		printf("%d ", arr[i]);
	}
	printf(" }\n");
}

File: Java/Classwork_2/src/telran/array/utils/ArrayTools.java
package telran.array.utils;

public class ArrayTools {

	
	public static void fillArray(int[] arr, int min, int max) {
		for (int i = 0; i < arr.length; i++) {
			arr[i] = (int) (min + Math.random() * (max + 1 - min));
			}
	}
	
	public static int sumArray(int[] arr) {
		int res = 0;
		for (int i = 0; i < arr.length; i++) {
			res += arr[i];
		}
		return res;
	}
	
	public static double avg(int[] arr) {
			return 1. * sumArray(arr) / arr.length;
	}
	
	public static void printArray(int[] arr) {
		for (int i = 0; i < arr.length; i++) {
			System.out.print(arr[i] + "\t");
		}
		System.out.println("\n===========");
	}

	public static void bubbleSort(int arr[]) {
		while (bubble(arr));
	}
	

	private static boolean bubble(int[] arr) {
		boolean flag = false;
		int i;
		for (i = 0; i < arr.length-1; i++) {
			if (arr[i] > arr[i+1]) {
				arr[i] = arr[i] + arr[i + 1];
				arr[i + 1] = arr[i] - arr[i + 1];
				arr[i] = arr[i] - arr[i + 1];
				flag = true;
			}
		}
		return flag;
	}
	
	
	public static int binarySearch(int arr[], int value) {
		int l = 0;
		int r = arr.length - 1;
		while (l <= r) {
			int mid = (l + r) / 2;
			if (value == arr[mid]) {
				return mid;
			}
			if (value < arr[mid]) {
				r = mid - 1;
			}
			if (value  > arr[mid]) {
				l = mid + 1;
			}
		}
		return -1;
	}
	
	public static void selectionSort(int arr[]) {
	int min;
	int index = 0;
	int i, j;
	for (i = 0; i < arr.length - 1; i++) {
		min = arr[i];
		for (j = i; j < arr.length; j++) {
			if (arr[j] < min) {
				min = arr[j];
				index = j;
			}
		}
		if (arr[i] > arr[index]) {
			int temp = arr[i];
			arr[i] = arr[index];
			arr[index] = temp;
		}
	}
	System.out.println(arr);
}
}